### PR TITLE
Scheduling stuff from 2017-11-13

### PR DESCRIPTION
**Pull request by pbrinkmeier**
I added the contents of the monday lecture and applied the style guide to the rest of the document.

### CODE REVIEW

File: betriebssysteme/06-scheduling.md [hunk 1]
```diff
@@ -1,194 +1,388 @@
...
+        - [Deadline class](#deadline-class)
+        - [Realtime class](#realtime-class)
+        - [CFS class](#cfs-class)
+        - [Idle class](#idle-class)
+    - [SMP scheduling in the Linux kernel](#smp-scheduling-in-the-linux-kernel)
 
 ## Dispatching
 
-The machine has `K` jobs ready to run, but only `N` CPUs with `K > N > 1`.
+The machine has K jobs ready to run, but only N CPUs where K > N > 1.
```
Comment (luk3b) [inline betriebssysteme/06-scheduling.md]: I think highlighting them as math/code symbols looks a bit better, else you easily overlook the `K` for example.

File: betriebssysteme/06-scheduling.md [hunk 2]
```diff
@@ -1,194 +1,388 @@
...
 
-#### Estimating the Length of Next CPU Burst
+t(n) = actual length of n-th CPU burst  
+τ(n + 1) = predicted value for the next CPU burst  
+τ(n + 1) = α * t(n) + (1 - α) * τ(n) where 0 ≤ α ≤ 1.
 
-Idea: use exponential averaging based on previous CPU bursts  
-`t(n)` = actual length of n-th CPU burst  
-`τ(n+1)` = predicted value for the next CPU burst  
-Define `τ(n+1) = α*t(n) + (1 - α)*τ(n)` with `0 <= α <= 1`
```
Comment (luk3b) [inline betriebssysteme/06-scheduling.md]: Why remove the backticks? Same as above, I think it is better to differentiate between text and math/code.

File: betriebssysteme/06-scheduling.md [hunk 3]
```diff
@@ -1,194 +1,388 @@
...
 
 Processes can be characterized as either:
-- **CPU-bound processes:** more time spent doing computations (very long but few CPU bursts)
-or
-- **I/O-bound processes:** more time spent doing I/O (many short CPU bursts)
+
+- **CPU-bound**: more time spent doing computations (very long but few CPU bursts)
+- **I/O-bound**: more time spent doing I/O (many short CPU bursts)
 
 ### Preemptive Shortest-Job-First (PSJF) Scheduling
```
Comment (luk3b) [inline betriebssysteme/06-scheduling.md]: The other headlines use "scheduling" in lower case, better use that here as well for consistency.

File: betriebssysteme/06-scheduling.md [hunk 4]
```diff
@@ -1,194 +1,388 @@
...
+Jobs in the additional queue are assigned a higher priority than normal jobs.
+After they have been run completely, they are put back in the normal queue.
+
+### (Strict) priority scheduling
+
+Each process is assigned an integer priority number.
+For each of those numbers, there is a queue.
+Allocate CPU time via round robin; start at the first queue (highest priority) and only execute jobs of lower priorities if there are no more important jobs (**strict priority scheduling**).
+
+![Priority queues](img/06-priority-scheduling.png)
```
Comment (luk3b) [inline betriebssysteme/06-scheduling.md]: Image is quite big, especially on mobile browsers.

File: betriebssysteme/06-scheduling.md [hunk 5]
```diff
@@ -1,194 +1,388 @@
...
+
+This way, it is possible that some processes never run, because their priority is too low (**starvation**).
+To counter this phenomenon, **aging** can be employed, i.e. processes move to a higher priority after waiting a while.
```
Comment (luk3b) [inline betriebssysteme/06-scheduling.md]: *ageing
http://www.english-for-students.com/ageing.html

File: betriebssysteme/06-scheduling.md [hunk 6]
```diff
@@ -1,194 +1,388 @@
...
+
+### Lottery scheduling
+
+Issue a total number n of **lottery tickets** to processes.
+The higher the priority of a process is, the more tickets it gets.
+
+Now for each time slice, generate a random number 0 ≤ r < n.
+Traverse the list of all processes, and give the slice to the first process where sum of ticket numbers is bigger than r.
+
+![Lottery scheduling example](img/06-lottery-example.png)
```
Comment (luk3b) [inline betriebssysteme/06-scheduling.md]: Quite big as well.

File: betriebssysteme/06-scheduling.md [hunk 7]
```diff
@@ -1,194 +1,388 @@
...
+They can be used for period realtime tasks — i.e. tasks that may be buffered, for example video encoding or decoding.
+
+#### Realtime class
+
+Realtime jobs are **POSIX real-time tasks**.
+They are assigned a priority ranging from 0 to 99 and there are two possible policies; SCHED_FIFO and SCHED_RR (round robin with a default time slice of 100 milliseconds).
+To run a realtime job on a Linux machine, execute the following (requires root privileges):
+
+```
+chrt --rr <priority> <task>
```
Comment (luk3b) [inline betriebssysteme/06-scheduling.md]: Just an idea: we could use a $ symbol for shell commands, while lines without the dollar symbol represent actual C code.
So this line would become
```$ chrt --rr <priority> <task>```

File: betriebssysteme/06-scheduling.md (hunk 1 above)
Comment (pbrinkmeier) [reply betriebssysteme/06-scheduling.md]: I think I will add a rule to the styleguide; use code tags for mathematical expressions, but keep special characters out of there (they tend to not be included in monospaced fonts).

Comment (pbrinkmeier) [reply betriebssysteme/06-scheduling.md]: (And then of course revise this lecture according to the new rule).

File: betriebssysteme/06-scheduling.md (hunk 2 above)
Comment (pbrinkmeier) [reply betriebssysteme/06-scheduling.md]: See above.

File: betriebssysteme/06-scheduling.md (hunk 3 above)
Comment (pbrinkmeier) [reply betriebssysteme/06-scheduling.md]: Didn't catch that one :P

File: betriebssysteme/06-scheduling.md (hunk 4 above)
Comment (pbrinkmeier) [reply betriebssysteme/06-scheduling.md]: I think we can't really do anything about that right now. IMO the images should stay the original size and more accurate image styling should only be done as soon as we compile the Markdown files ourselves.

File: betriebssysteme/06-scheduling.md (hunk 5 above)
Comment (pbrinkmeier) [reply betriebssysteme/06-scheduling.md]: Whoops :D

File: betriebssysteme/06-scheduling.md (hunk 7 above)
Comment (pbrinkmeier) [reply betriebssysteme/06-scheduling.md]: Ok, can do

File: betriebssysteme/06-scheduling.md (hunk 4 above)
Comment (luk3b) [reply betriebssysteme/06-scheduling.md]: Hmm I don't really want to mix Markdown and HTML, however the images are very wide (even on GitHub desktop).
How do you want to style them later? That will really depend on what format we'll compile it into.

**Comment by pbrinkmeier** [comment]
I just thought we'd set `float: right` on all img tags in HTML.

2017-11-22 10:03 GMT+01:00 Luke Brocke <notifications@github.com>:

File: betriebssysteme/06-scheduling.md (hunk 4 above)
Comment (luk3b) [reply betriebssysteme/06-scheduling.md]: > I just thought we'd set `float: right` on all img tags in HTML.
We could do that, yes, however that assumes we're exporting to HTML. I don't know whether using HTML will break anything when exporting to LaTeX for example.

**Comment by pbrinkmeier** [comment]
Pls review

**Pull request merged**